pilot/pkg/status/reporter.go

```diff
@@ -0,0 +1,289 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	v2 "istio.io/istio/pilot/pkg/proxy/envoy/v2"
+
+	"github.com/pkg/errors"
+
+	"k8s.io/client-go/rest"
+
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/utils/clock"
+
+	"istio.io/istio/pilot/pkg/model"
+)
+
+func NewIstioContext(stop <-chan struct{}) context.Context {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-stop
+		cancel()
+	}()
+	return ctx
+}
+
+type inProgressEntry struct {
+	// the resource, including resourceVersion, we are currently tracking
+	Resource
+	// the number of reports we have written with this resource at 100%
+	completedIterations int
+}
+
+type Reporter struct {
+	mu                  sync.RWMutex
+	status              map[string]string
+	reverseStatus       map[string][]string
+	dirty               bool
+	inProgressResources map[string]*inProgressEntry
+	client              v1.ConfigMapInterface
+	cm                  *corev1.ConfigMap
+	UpdateInterval      time.Duration
+	PodName             string
+	clock               clock.Clock
+	store               model.ConfigStore
+}
+
+const labelKey = "internal.istio.io/distribution-report"
+const dataField = "distribution-report"
+
+// Starts the reporter, which watches dataplane ack's and resource changes so that it can update status leader
+// with distribution information
+func (r *Reporter) Start(restConfig *rest.Config, namespace string, store model.ConfigStore, stop <-chan struct{}) {
+	scope.Info("Starting status follower controller")
+	ctx := NewIstioContext(stop)
+	if r.clock == nil {
+		r.clock = clock.RealClock{}
+	}
+	r.store = store
+	// default UpdateInterval
+	if r.UpdateInterval == 0 {
+		r.UpdateInterval = 500 * time.Millisecond
+	}
+	clientSet, err := kubernetes.NewForConfig(restConfig)
```

Comment:
Please please stop creating new client sets and clients ! Every package has its own code. 

We have a client and rest config created by istiod at startup. Just pass it around.

Reply:
Can you tell me where to find the other clientset?  I asked in this PR before but no one answered...

Reply:
bootstrap package should have some - look how the security or validation are initialized.

Reply:
found it, and refactored to use it rather than creating a new one.